Masking proxy credentials from URL when displayed in system info

dani-docker opened the pull request:
Signed-off-by: Dani Louca <dani.louca@docker.com>

If dockerd is configured to use a proxy https://docs.docker.com/config/daemon/systemd/#httphttps-proxy and if the proxy URL is configured for basic authentication, the `SystemInfo` (ex: `docker info`) shows the full proxy URL, including the credentials.
 
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
This PR masks out the proxy username / password while building the info struct.
To make the solution robust, I used the golang `url` package to identify the credentials, replace them with asterisks, and rebuild the URL string.
I opted not to use basic strings replacement as it will require to address a bunch of corner cases which makes the solution more complex .

**- How I did it**
- convert the proxy URL string into a URL object
- overwrite the URL object's credentials with asterisks
- convert the URL to string, (this will result of an encoded string)
- decode the encoded string , if the decoder fails, simply returns the encoded string

**- How to verify it**
- configure your docker to use the following proxy URL `http://USER:PASSWORD@proxy.example.com:80/` by following https://docs.docker.com/config/daemon/systemd/#httphttps-proxy
- run `docker info | grep Proxy`
**Expected Results:**
`Http Proxy: http://*********:*********@proxy.example.com:80/`
**Actual Result:**
`Http Proxy: http://USER:PASSWORD@proxy.example.com:80/`
**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

Commit:
Masking credentials from proxy URL

Signed-off-by: Dani Louca <dani.louca@docker.com>

dani-docker commented:
cc: @thaJeztah

thaJeztah reviewed (changes requested):
Thanks! Left some comments inline.

While I can see the reason for masking this information (especially in the output of `docker info`), I have some concerns; this change could be a breaking change for those that use this information (for whatever reason; perhaps automation, or verifying daemon configuration). This information was added a long time ago (since https://github.com/moby/moby/pull/11125), so there's definitely a chance that people use this.

Masking this information in just `docker info` / the `/info` endpoint _limits_ the exposure, but does not prevent this information from leaking in other ways (this config is obtained from their corresponding `HTTP(S)_PROXY` environment variables); also taking into account that someone with access to the API already effectively is "root" on the host.

Perhaps not an issue, but worth to double check if we anticipate this being a breaking change.

thaJeztah commented:
Oh, we should update the swagger file as well, and update the description for these fields, to describe that credentials are masked when using the endpoint; https://github.com/moby/moby/blob/7bfec8cd80c5e2ce548cd7ccf63a35bf60d802d3/api/swagger.yaml#L3760-L3784

Also may warrant a mention in the API changelog https://github.com/moby/moby/blob/7bfec8cd80c5e2ce548cd7ccf63a35bf60d802d3/docs/api/version-history.md (although it's not versioned, so not sure how to best describe it there)

dani-docker commented on daemon/info.go:
```diff
@@ -245,3 +246,19 @@ func operatingSystem() string {
+	decodedURL, err := url.QueryUnescape(maskedURL)
+	// return encoded maskedURL in case of an error
+	if err != nil {
+		logrus.Errorf("Failed to decode masked URL, returning encoded masked URL: %v", err)
```
good point, I will change that

dani-docker commented on daemon/info.go:
```diff
@@ -245,3 +246,19 @@ func operatingSystem() string {
+	}
+	parsedURL.User = url.UserPassword("*********", "*********")
+	maskedURL := parsedURL.String()
+	decodedURL, err := url.QueryUnescape(maskedURL)
```
@thaJeztah 
Right, I would assume that it's not common to for user to enter a URL encoded string, but if this is common, I can look into adding some logic to check if a string is URL encoded before decoding.

It's not limited to `xxxxxx` or `*****`, as the encoding decoding also applies to the rest of the URL, as the `.String()` func always return an encoded string.

thaJeztah replied:
True, but the other parts would be where URL-encoding would be needed from the start (probably in the original value already), whereas the `**` was what we added, and want to have presented as-is (without encoding)

dani-docker replied:
@thaJeztah 
you are correct, the `String()` does not change the end of the URL.
replacing with `xxxxx` will make things much simpler

dani-docker commented:
> Oh, we should update the swagger file as well, and update the description for these fields, to describe that credentials are masked when using the endpoint;
> 
> [moby/api/swagger.yaml](https://github.com/moby/moby/blob/7bfec8cd80c5e2ce548cd7ccf63a35bf60d802d3/api/swagger.yaml#L3760-L3784)
> 
> Lines 3760 to 3784 in [7bfec8c](/moby/moby/commit/7bfec8cd80c5e2ce548cd7ccf63a35bf60d802d3)
> 
>        HttpProxy: 
>          description: | 
>            HTTP-proxy configured for the daemon. This value is obtained from the 
>            [`HTTP_PROXY`](https://www.gnu.org/software/wget/manual/html_node/Proxies.html) environment variable. 
>   
>            Containers do not automatically inherit this configuration. 
>          type: "string" 
>          example: "http://user:pass@proxy.corp.example.com:8080" 
>        HttpsProxy: 
>          description: | 
>            HTTPS-proxy configured for the daemon. This value is obtained from the 
>            [`HTTPS_PROXY`](https://www.gnu.org/software/wget/manual/html_node/Proxies.html) environment variable. 
>   
>            Containers do not automatically inherit this configuration. 
>          type: "string" 
>          example: "https://user:pass@proxy.corp.example.com:4443" 
>        NoProxy: 
>          description: | 
>            Comma-separated list of domain extensions for which no proxy should be 
>            used. This value is obtained from the [`NO_PROXY`](https://www.gnu.org/software/wget/manual/html_node/Proxies.html) 
>            environment variable. 
>   
>            Containers do not automatically inherit this configuration. 
>          type: "string" 
>          example: "*.local, 169.254/16" 
> Also may warrant a mention in the API changelog https://github.com/moby/moby/blob/7bfec8cd80c5e2ce548cd7ccf63a35bf60d802d3/docs/api/version-history.md (although it's not versioned, so not sure how to best describe it there)

yes, good catch, I will update the swagger

dani-docker commented:
I searched the moby code base and I don't see any place that reads the `HTTPProxy` or `HTTPSProxy` straight from the info struct.
I do share your concerns that some 3rd party code might be reading this info from the API and if this is the case, then we'll break their integration; but even if they do, I think it's still not correct design that we share the credentials through the `info` endpoint.

dani-docker commented:
@thaJeztah 
Changed the implementation as suggested and added couple of extra unit tests.
Note: I was under the impression that `String()` encodes the rest of the URL, not only the credentials part, but running few tests showed that it's not the case;  by using `xxxxx` instead of `*****` , we no longer need to worry about decoding/encoding the result of `String()`

Another note, while testing different URLs variations; 
I noticed that the `url` method `parse` will fail to parse out credentials that are not escaped. 
for example running `parse` on this url `http://USER@docker:pa?sword@proxy.example.com:80/hello world` , will result of no password or username set, this will cause the masking URL to look like this `http://xxxxx:xxxxx@docker:pa?sword@proxy.example.com:80/hello world`

The expectation here is that user should encode those type of strings when they configure HTTP(s) Proxy, otherwise, their dial will also fail as the core logic uses the same `parse` function

dani-docker commented:
Also updated the swagger but not sure how to deal with the history file without having an API version

dani-docker commented on api/swagger.yaml:
```diff
@@ -3760,19 +3760,19 @@ definitions:
         description: |
           HTTP-proxy configured for the daemon. This value is obtained from the
-          [`HTTP_PROXY`](https://www.gnu.org/software/wget/manual/html_node/Proxies.html) environment variable.
+          [`HTTP_PROXY`](https://www.gnu.org/software/wget/manual/html_node/Proxies.html) environment variable and credentials are masked.
```
Thanks @thaJeztah , changes are made

dani-docker replied:
thanks! change is made

dani-docker commented on api/swagger.yaml:
```diff
@@ -3760,19 +3760,19 @@ definitions:
         description: |
           HTTPS-proxy configured for the daemon. This value is obtained from the
-          [`HTTPS_PROXY`](https://www.gnu.org/software/wget/manual/html_node/Proxies.html) environment variable.
+          [`HTTPS_PROXY`](https://www.gnu.org/software/wget/manual/html_node/Proxies.html) environment variable and credentials are masked.
```
thanks! change is made

thaJeztah reviewed (approved):
LGTM, thanks!

thaJeztah commented:
It's green! merging
